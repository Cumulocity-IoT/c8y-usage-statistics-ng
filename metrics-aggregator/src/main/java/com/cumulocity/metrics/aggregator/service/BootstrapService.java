package com.cumulocity.metrics.aggregator.service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */

@Service
public class BootstrapService {

    @Autowired
    DeviceMetricsAggregationService deviceMetricsAggregationService;
 
    @Autowired
    StartupUtils utils;

    private static final Logger log = LoggerFactory.getLogger(BootstrapService.class);

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        utils.setCurrentTenant(event.getCredentials().getTenant());
        utils.getTenants();
        deviceMetricsAggregationService.createDailyDeviceStatistics();
	}

}
