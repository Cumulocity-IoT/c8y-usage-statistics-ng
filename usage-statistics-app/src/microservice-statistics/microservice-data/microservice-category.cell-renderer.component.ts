import { Component } from "@angular/core";
import { CellRendererContext } from "@c8y/ngx-components";

@Component({
  template: `
    <span>{{ categoryName }}</span>
  `,
})
export class MicroserviceCategoryCellRendererComponent {
  categoryName: string;

  constructor(
    public context: CellRendererContext,
  ) {
    // this will be updated later
    const map = {
      "databroker-agent-server": "Product Services",
      "zementis-large": "Custom Microservices",
      "billing-aggregation": "Custom Microservices",
      "billing-fetcher": "Custom Microservices"
    }

    this.categoryName = map[context.value] || "Custom Microservices";
  }
}

