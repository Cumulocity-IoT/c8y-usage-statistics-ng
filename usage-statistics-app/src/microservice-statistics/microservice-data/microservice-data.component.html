<c8y-title>
    {{ "Microservice Data" | translate }}
  </c8y-title>
  
  <statistics-action-bar [feature]="feature"></statistics-action-bar>
  <ng-container>
    <div class="container-fluid">
      <table class="table">
        <!-- <colgroup>
          <col width="20px">
          <col width="50%">
        </colgroup> -->
        <thead>
          <tr>
            <th>Total avg Memory</th>
            <th>Total avg CPU</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td >{{avgMEM}}</td>
            <td >{{avgCPU}}</td>
          </tr>
        </tbody>
      </table>
    </div>
    
    </ng-container>
  <div class="content-fullpage d-flex d-col shadow1">
      <c8y-data-grid
      [loadMoreItemsLabel]="loadMoreItemsLabel"
      [loadingItemsLabel]="loadingItemsLabel"
      [columns]="columns"
      [rows]="microserviceData"
      [pagination]="pagination"
      [infiniteScroll]="infiniteScroll"
      [actionControls]="actionControls"
      [showSearch]="showSearch"
      class="d-contents"
    >
      <div class="c8y-empty-state" *ngIf="isLoading">
        <h1 c8yIcon="search"></h1>
        <div>
          <p>
            <strong *ngIf="isLoading" translate>{{loadingItemsLabel}} </strong>
          </p>
        </div>
      </div>
      <div class="c8y-empty-state" *ngIf="!isLoading && !(microserviceData && microserviceData.length > 0)">
        <h1 c8yIcon="search"></h1>
        <div>
          <p>
            <strong>{{ "Microservice details are not available." | translate }}</strong>
          </p>
        </div>
      </div>
      <ng-container *ngFor="let column of columns">
        <c8y-column [name]="column.name"></c8y-column>
      </ng-container>
    </c8y-data-grid>
  </div>