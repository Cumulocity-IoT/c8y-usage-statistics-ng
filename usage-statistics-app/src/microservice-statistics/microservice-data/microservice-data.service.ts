import { Injectable } from "@angular/core";
import { FetchClient, IFetchOptions } from "@c8y/client";
import { gettext } from "@c8y/ngx-components";

export const MEMORY: string = 'memory';
export const NAME: string = 'name';
export const CPU: string = 'cpu';
export const CAUSE: string = 'cause';
export const CATEGORY: string = 'category'
export const COLUMN_FIELDS = {
  MICROSERVICE: gettext('Microservice'),
  MEMORY_TOTAL: gettext('Memory (MiB)'),
  MEMORY_AVG: gettext('Daily Avg Memory (GiB)'),
  CPU_TOTAL: gettext('CPU (mCPU)'),
  CPU_AVG: gettext('Daily Avg CPU (CPUs)'),
  CAUSE: gettext('Source')
}

@Injectable({
  providedIn: "root",
})
export class MicroserviceDataService {}