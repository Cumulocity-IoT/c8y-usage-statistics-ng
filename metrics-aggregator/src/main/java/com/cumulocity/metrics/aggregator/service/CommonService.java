package com.cumulocity.metrics.aggregator.service;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

import com.cumulocity.metrics.aggregator.model.microservice.Tenants;
import com.cumulocity.metrics.aggregator.model.microservice.Tenants.Tenant;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.sdk.client.RestConnector;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */
@Service
public class CommonService {

    @Autowired
    RestConnector restConnector;
    

    @Autowired
	MicroserviceSubscriptionsService subscriptionsService;

    private static final Logger log = LoggerFactory.getLogger(CommonService.class);

    private String currentTenant = ""; 
    private List<String> tenantList;

    public String getCurrentTenant() {
        return currentTenant;
    }

    public List<String> getTenantList() {
        return tenantList;
    }

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        this.currentTenant = event.getCredentials().getTenant();
		log.info("CurrentTenant: " + this.currentTenant);

        this.tenantList = subscriptionsService.callForTenant(this.currentTenant, this::getSubTenants).get();
        this.tenantList.add(currentTenant);
        log.info(tenantList.toString());
	}

    private Optional<List<String>> getSubTenants(){
        Tenants tenants = restConnector.get(
                "/tenant/tenants?pageSize=2000&withTotalPages=true&withApps=false",
                CumulocityMediaType.APPLICATION_JSON_TYPE,Tenants.class);
            ArrayList<String> tenantIds = new ArrayList<String>();
            for (Tenant tenant : tenants.getTenants()){
                tenantIds.add(tenant.getId());
            }
            return Optional.of(tenantIds);
    }

}
