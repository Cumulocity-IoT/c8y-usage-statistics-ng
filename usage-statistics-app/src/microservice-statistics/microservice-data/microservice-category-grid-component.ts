import { Type } from '@angular/core';
import { Column, ColumnDataType, gettext } from '@c8y/ngx-components';
import { MicroserviceCategoryCellRendererComponent } from './microservice-category.cell-renderer.component';
import { CATEGORY, NAME } from './microservice-data.service';

export class MicroserviceCategoryGridColumn implements Column {
  name: string;
  path?: string;
  header?: string;
  dataType?: ColumnDataType;

  visible?: boolean;
  positionFixed?: boolean;
  gridTrackSize?: string;

  cellRendererComponent?: Type<any>;

  sortable?: boolean;

  filterable?: boolean;

  constructor() {
    this.name = CATEGORY;
    this.path = NAME; // we need the context to be name inorder to map the name against product category.
    this.header = gettext('Category');
    this.cellRendererComponent = MicroserviceCategoryCellRendererComponent;

    this.filterable = true;
    this.sortable = true;
  }
}