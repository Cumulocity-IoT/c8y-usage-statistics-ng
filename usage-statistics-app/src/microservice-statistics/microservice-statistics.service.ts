import { Injectable } from "@angular/core";
import { DateTimePickerComponent, gettext } from "@c8y/ngx-components";

import { Api, MicroserviceCategory, MicroserviceConfigurationService, ProductCategory, PropertyName } from './microservice-configuration/microservice-configuration.service';
import { CommonService, DATE_FORMAT_MONTH } from "../common.service";
import { FetchClient } from "@c8y/client";

const moment = require('moment');
export const CLASS_COLORS = ['#024B7A', '#44A5C2', '#FFAE49', '#2B8A3E', '#58508D', '#9C1086', '#E0000E', '#E67700', '#004225', '#592720'];
export const DATE_FORMAT_DAY = 'YYYY-MM-DD';
export interface TenantSummaryResources {
    memory: number,
    name: string,
    cpu: number,
    cause: string
}

export interface MonthlyMicroserviceProdCategoryMap extends TenantSummaryResources {
    [PropertyName.Microservice]: string,
    avgCpu?: string,
    avgMemory?: string
}

@Injectable({
    providedIn: 'root'
})
export class MicroserviceStatisticsService {
    public microserviceStatisticsDataStore = {
        response: <MonthlyMicroserviceProdCategoryMap[]>[],
        date: <Date>new Date(),
        avgCPU: "",
        avgMEM: "" 
    }

    constructor(
        private microserviceConfigurationService: MicroserviceConfigurationService,
        private commonService: CommonService,
        private client: FetchClient,
    ) { }

   
    
    getProductService(): string[]{
        return ["actility",
                "administration",
                "advanced-software-mgmt",
                "apama-ctrl-1c-4g",
                "apama-ctrl-250mc-1g",
                "apama-ctrl-2c-8g",
                "apama-ctrl-smartrulesmt",
                "apama-ctrl-starter",
                "apama-oeeapp", "apama-subscription-mgr",
                "billwerk-agent-server",
                "cellid", "c8y-community-utils",
                "c8ydata", "cloud-remote-access",
                "cockpit",
                "cockpitbeta",
                "connectivity-agent-server",
                "core",
                "databroker-agent-server",
                "datahub",
                "DataHub",
                "device-counter-exporter",
                "devicemanagement",
                "device-simulator",
                "device-statistics-week",
                "digital-twin-manager",
                "dtm-ms",
                "dtm",
                "feature-cep-custom-rules",
                "feature-fieldbus4",
                "feature-microservice-hosting",
                "feature-user-hierarchy",
                "impact",
                "loriot-agent",
                "management", "metadata-collector",
                "OEE", "onnx",
                "opcua-mgmt-service",
                "oee-bundle",
                "public-options",
                "report-agent",
                "repository-connect",
                "sigfox-agent",
                "smartrule",
                "sms-gateway",
                "sslmanagement",
                "stratos-client",
                "Streaming Analytics",
                "tenant-cleanup",
                "tenant-data-exporter",
                "tenant-recovery",
                "zementis-small",
                "mlw",
                "nyoka",
                "cep",
                "apama-ctrl-24c-16g",
                "apama-ctrl-4c-8g",
                "apama-ctrl-025c-1g",
                "apama-ctrl-mt-4c-16g",
                "metrics-aggregator",
                "device-statistics-week"]
      }

    async getMonthlyMicroserviceProdCategoryMap(selectedDate: Date): Promise<MonthlyMicroserviceProdCategoryMap[]> {
        //const productCategories: ProductCategory[] = await this.microserviceConfigurationService.getCategories(Api.Product, false)        
        const defaultProductCategory = "Custom Microservice"
        //const microserviceCategories = await this.microserviceConfigurationService.getCategories(Api.Microservice, false)
        const tenantSummaryResources = await this.getTenantSummaryResources(selectedDate)
        const response: MonthlyMicroserviceProdCategoryMap[] = [];
        const numberOfDaysInMonth = moment(selectedDate, DATE_FORMAT_MONTH).daysInMonth();
        let totalCPU:number = 0;
        let totalMEM:number= 0;

        tenantSummaryResources.forEach((elem: TenantSummaryResources) => {
            const microserviceName = elem.name
            if (!this.getProductService().includes( microserviceName)){
                totalCPU = totalCPU + elem.cpu;
                totalMEM = totalMEM + elem.memory;
                response.push({
                    ...elem,
                    microserviceName: microserviceName,
                    avgCpu: (elem.cpu / (1000 * numberOfDaysInMonth)).toFixed(2),
                    avgMemory: (elem.memory / (4294.97 * numberOfDaysInMonth)).toFixed(2)
                })
            }
        })
        this.microserviceStatisticsDataStore.avgCPU = (totalCPU / (1000 * numberOfDaysInMonth)).toFixed(2);
        this.microserviceStatisticsDataStore.avgMEM = (totalMEM  / (4294.97 * numberOfDaysInMonth)).toFixed(2);

        this.microserviceStatisticsDataStore.response = response
        this.microserviceStatisticsDataStore.date = selectedDate
        return response
    }

    private getProductCategory(microserviceCategories: MicroserviceCategory[], defaultProductCategory, microserviceName: string) {
        const microserviceMapping = microserviceCategories.find((elem: MicroserviceCategory) => elem.microserviceName.trim() === microserviceName.trim())
        return microserviceMapping ? microserviceMapping.productCategory : defaultProductCategory;
    }


    private async getTenantSummaryResources(selectedDate: Date): Promise<TenantSummaryResources[]> {
        const data = await this.commonService.getCurrentTenantSummary(selectedDate)
        if (data && data.resources && data.resources.usedBy) {
            return this.getAggregatedTenantSummary(data.resources.usedBy) as TenantSummaryResources[];
        }
        else {
            throw { message: gettext('Microservice data is not available') }
        }
    }

    private getAggregatedTenantSummary(data: TenantSummaryResources[]): TenantSummaryResources[] | null {
        const map = {}
        data.forEach(elem => {
            if (map[elem.name]) {
                map[elem.name].cpu += elem.cpu
                map[elem.name].memory += elem.memory
                if (!map[elem.name].cause.includes(elem.cause)) {
                    map[elem.name].cause.push(elem.cause)
                }
            }
            else {
                map[elem.name] = {
                    name: elem.name,
                    cpu: elem.cpu,
                    memory: elem.memory,
                    cause: [elem.cause]
                }
            }
        })

        const result = Object.values(map)
        result.forEach((elem: any) => {
            elem.cause = elem.cause.join(", ")
        })
        return result as TenantSummaryResources[]
    }

    public async getMonthlyMicroserviceAggregation(date: Date):Promise<any> {

        var dateTo: Date;
        dateTo= new Date(date.getFullYear(), date.getMonth() + 1, 0);
       
        console.log("dateTo: " + dateTo)

        try {
    
          const options = {
            method: 'GET'
          };
          const response = await this.client.fetch(
            `/service/metrics-aggregator/microservices/?dateFrom=${moment(date).format(DATE_FORMAT_DAY)}&dateTo=${moment(dateTo).format(DATE_FORMAT_DAY)}`
            , options);
          return (await response).json();
        } catch (err) {
          console.log(err);
        }
      }

}