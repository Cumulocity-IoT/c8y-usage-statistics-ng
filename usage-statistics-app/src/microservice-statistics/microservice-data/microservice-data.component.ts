import { Component, ElementRef, OnDestroy, OnInit } from '@angular/core';
import { gettext, DisplayOptions, Pagination, Column, CellRendererContext } from '@c8y/ngx-components';
import { Subscription } from 'rxjs';
import { CommonService, FeatureList } from '../../common.service';
import { MonthPickerService } from '../../utitities/statistics-action-bar/month-picker/month-picker.service';
import { MicroserviceStatisticsService } from '../microservice-statistics.service';
import { COLUMN_FIELDS } from './microservice-data.service';

const moment = require('moment');



@Component({
  selector: 'app-microservice-data',
  templateUrl: './microservice-data.component.html',
  styleUrls: ['./microservice-data.component.css']
})
export class MicroserviceDataComponent implements OnInit, OnDestroy {
  feature = FeatureList.MicroserviceStatistics;
  monthChangedSubscription: Subscription;
  tenantChangedSubscription: Subscription;
  microserviceData: any[];
  loadMoreItemsLabel: string = gettext('Load more microservice details');
  loadingItemsLabel: string = gettext('Loading microservice details...');
  displayOptions: DisplayOptions = {
    bordered: false,
    striped: true,
    filter: true,
    gridHeader: true,
    hover: true
  };
  isLoading = true;

  MEMORY: string = 'memory';
  NAME: string = 'microserviceName';
  CPU: string = 'cpu';
  CAUSE: string = 'cause';
  CATEGORY: string = 'productCategory';
  AVG_CPU: string = 'avgCpu';
  AVG_MEMORY: string = 'avgMemory';

  columns: Column[] = this.getColumns();
  pagination: Pagination = this.getPagination();

  constructor(
    private monthPickerService: MonthPickerService,
    private microserviceStatisticsService: MicroserviceStatisticsService,
    private element: ElementRef,
    private commonService: CommonService
  ) {
  }

  ngOnInit(): void {
    this.monthChangedSubscription = this.monthPickerService.dateChanged.subscribe((selectedDate: Date) => {
      this.getMicroserviceData(selectedDate);
    });
    this.tenantChangedSubscription = this.commonService.tenantChanged.subscribe((tenantId: string) =>{
      this.getMicroserviceData();
    })
    this.getMicroserviceData();
    this.isLoading = false;
  }

  ngAfterViewInit() {
    const btns = this.element.nativeElement.querySelectorAll('c8y-data-grid .btnbar-btn');
    btns.forEach((btn: HTMLElement) => {
      if (btn.querySelector('i[c8yIcon=refresh]')) btn.style.display = 'none';
    });
  }

  getColumns(): Column[] {
    return [
      {
        name: this.NAME,
        header: COLUMN_FIELDS.MICROSERVICE,
        path: this.NAME,
        filterable: true,
        sortable: true
      },
      {
        name: this.CATEGORY,
        header: COLUMN_FIELDS.PROD_CATEGORY,
        path: this.CATEGORY,
        filterable: true,
        sortable: true
      },
      {
        name: this.MEMORY,
        header: COLUMN_FIELDS.MEMORY_TOTAL,
        path: this.MEMORY,
        filterable: true,
        sortable: true,
        visible: false,
        headerCellRendererComponent: MemoryCpuHeaderComponent
      },
      {
        name: this.AVG_MEMORY,
        header: COLUMN_FIELDS.MEMORY_AVG,
        path: this.AVG_MEMORY,
        filterable: true,
        sortable: true,
        headerCellRendererComponent: MemoryCpuHeaderComponent
      },
      {
        name: this.CPU,
        header: COLUMN_FIELDS.CPU_TOTAL,
        path: this.CPU,
        filterable: true,
        sortable: true,
        visible: false,
        headerCellRendererComponent: MemoryCpuHeaderComponent
      },
      {
        name: this.AVG_CPU,
        header: COLUMN_FIELDS.CPU_AVG,
        path: this.AVG_CPU,
        filterable: true,
        sortable: true,
        headerCellRendererComponent: MemoryCpuHeaderComponent
      },
      {
        name: this.CAUSE,
        header: COLUMN_FIELDS.CAUSE,
        path: this.CAUSE,
        filterable: true,
        sortable: true
      }
    ];
  }

  getPagination(): Pagination {
    return {
      pageSize: 10,
      currentPage: 1
    };
  }

  async getMicroserviceData(selectedDate = this.monthPickerService.selectedDate) {
    try {
      this.microserviceData = await this.microserviceStatisticsService.getMonthlyMicroserviceProdCategoryMap(selectedDate);
    }
    catch (error) {
      this.commonService.microserviceUnavailableAlert(error)
    }
  }

  ngOnDestroy(): void {
    if (this.monthChangedSubscription) {
      this.monthChangedSubscription.unsubscribe()
    }
    if (this.tenantChangedSubscription) {
      this.tenantChangedSubscription.unsubscribe()
    }
  }
}


@Component({
  template: `  
      {{ context?.value }}
      <button type="button" c8yicon="question-circle-o" class="btn btn-link dlt-c8y-icon-question-circle-o" style="text-decoration: none"
      title="{{ title }}">
      </button>     
  `
})
export class MemoryCpuHeaderComponent {
  title: string
  constructor(public context: CellRendererContext) {
    if (context?.value === COLUMN_FIELDS.MEMORY_AVG) {
      this.title = gettext('Average memory usage / day (GiB)');
    }
    else if (context?.value === COLUMN_FIELDS.MEMORY_TOTAL) {
      this.title = gettext('Total megabytes used in the month, calculated daily (MiB)');
    }
    else if (context?.value === COLUMN_FIELDS.CPU_AVG) {
      this.title = gettext('Average CPU usage / day (CPUs)');
    }
    else if (context?.value === COLUMN_FIELDS.CPU_TOTAL) {
      this.title = gettext('Total millicores used in the month, calculated daily (mCPU)');
    }
  }
}
