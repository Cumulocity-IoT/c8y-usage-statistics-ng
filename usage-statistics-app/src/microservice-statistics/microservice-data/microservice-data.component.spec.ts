import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { CoreModule } from '@c8y/ngx-components';
import { MicroserviceStatisticsService } from '../microservice-statistics.service';
import { MicroserviceDataComponent } from './microservice-data.component';

describe('Test MicroserviceDataComponent', () => {
  let component: MicroserviceDataComponent;
  let fixture: ComponentFixture<MicroserviceDataComponent>;
  let microserviceStatisticsService: MicroserviceStatisticsService;

  const microserviceStatisticsServiceMock = {
    getMonthlyMicroserviceProdCategoryMap: jest.fn(),
}

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot(), BrowserAnimationsModule],
      declarations: [MicroserviceDataComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: MicroserviceStatisticsService, useValue: microserviceStatisticsServiceMock },
    ]
    });
    fixture = TestBed.createComponent(MicroserviceDataComponent);
    component = fixture.componentInstance;
    microserviceStatisticsService = TestBed.inject(MicroserviceStatisticsService);
  });

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should call getMonthlyMicroserviceProdCategoryMap from service', async () => {
    fixture.detectChanges();
    const getMonthlyMicroserviceProdCategoryMapSpy = jest.spyOn(microserviceStatisticsService, 'getMonthlyMicroserviceProdCategoryMap');
    expect(getMonthlyMicroserviceProdCategoryMapSpy).toHaveBeenCalled();
  });

  test('should give correct device data response', async() => {
    fixture.detectChanges();
    let today = new Date("01 Jan 2023");
    const selectedDate = new Date(today.setMonth(today.getMonth() - 1));
    const mockResponse = [
        {
            memory: 31067,
            name: "databroker-agent-server",
            cause: "Subscribed",
            cpu: 28931,
            productCategory: "Custom Microservices",
            microserviceName: "databroker-agent-server"
        },
        {
            memory: 335018,
            name: "zementis-large",
            cause: "Subscribed",
            cpu: 167509,
            productCategory: "Custom Microservices",
            microserviceName: "zementis-large"
        },
        {
            memory: 15036,
            name: "datahub",
            cause: "Subscribed",
            cpu: 3500,
            productCategory: "Product Services",
            microserviceName: "datahub"
        },
        {
            memory: 111671,
            name: "nyoka",
            cause: "Subscribed",
            cpu: 55836,
            productCategory: "Machine Learning",
            microserviceName: "nyoka"
        }

    ];
    jest.spyOn(microserviceStatisticsService, 'getMonthlyMicroserviceProdCategoryMap').mockImplementation(()=>Promise.resolve(mockResponse))
    await component.getMicroserviceData(selectedDate);
    expect(component.microserviceData).toEqual(mockResponse);
  });
});