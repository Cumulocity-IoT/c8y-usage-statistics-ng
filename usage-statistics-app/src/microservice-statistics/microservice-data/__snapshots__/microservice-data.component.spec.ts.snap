// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MicroserviceDataComponent should load expected component 1`] = `
<app-microservice-data
  AVG_CPU={[Function String]}
  AVG_MEMORY={[Function String]}
  CATEGORY={[Function String]}
  CAUSE={[Function String]}
  CPU={[Function String]}
  MEMORY={[Function String]}
  NAME={[Function String]}
  columns={[Function Array]}
  commonService={[Function CommonService]}
  displayOptions={[Function Object]}
  element={[Function ElementRef]}
  feature={[Function String]}
  isLoading={[Function Boolean]}
  loadMoreItemsLabel={[Function String]}
  loadingItemsLabel={[Function String]}
  microserviceStatisticsService={[Function Object]}
  monthPickerService={[Function MonthPickerService]}
  pagination={[Function Object]}
>
  <c8y-title>
  </c8y-title><statistics-action-bar /><div
    class="content-fullpage d-flex d-col shadow1"
  >
    <c8y-data-grid
      class="d-contents"
    >
      <div
        class="table-data-grid-scroll"
        data-cy="c8y-data-grid--table-data-grid-scroll"
      >
        <table
          cdk-table=""
          class="cdk-table table table-hover table-filtered-sorted table-data-grid large-padding"
          data-cy="c8y-data-grid--table"
          role="table"
        >
        </table>
      </div>
    </c8y-data-grid>
  </div>
</app-microservice-data>
`;
